File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/typescript/evm/swaps/getSwapPrice.ts
// Usage: pnpm tsx evm/swaps/getSwapPrice.ts

/**
 * This example demonstrates how to get swap pricing information without executing a swap.
 * 
 * Use getSwapPrice when you need:
 * - Quick price estimates for display purposes
 * - Exchange rate calculations for UI components  
 * - Liquidity checks before attempting a swap
 * - Price comparisons across different token pairs
 * 
 * Key differences from createSwapQuote:
 * - getSwapPrice: Returns pricing data only, no transaction details
 * - getSwapPrice: Can be queried more frequently, but data may be less precise
 * - createSwapQuote: Returns full transaction data ready for execution
 * - createSwapQuote: May reserve funds on-chain (represents a soft commitment to swap)
 * - createSwapQuote: More strictly rate-limited to prevent abuse
 * 
 * getSwapPrice is ideal for:
 * - Price calculators and converters
 * - Displaying "you will receive approximately X tokens"
 * - Real-time price updates in your UI
 * - Checking if a swap route exists before showing swap UI
 * 
 * Note: getSwapPrice provides estimates only. Actual swap execution may have
 * slight variations due to market movements between pricing and execution.
 */

import { CdpClient } from "@coinbase/cdp-sdk";
import { formatUnits, parseEther, type Address } from "viem";
import "dotenv/config";

const cdp = new CdpClient();

// Token definitions for the example
const TOKENS = {
  WETH: {
    address: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    symbol: "WETH",
    decimals: 18,
  },
  USDC: {
    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    symbol: "USDC",
    decimals: 6,
  },
};

async function main() {
  // Get or create an account to use for the swap
  const ownerAccount = await cdp.evm.getOrCreateAccount({ name: "SwapAccount" });
  console.log(`Using account: ${ownerAccount.address}`);

  try {
    // Define the tokens we're working with
    const fromToken = TOKENS.WETH;
    const toToken = TOKENS.USDC;
    
    // Get a price for swapping WETH to USDC on Ethereum mainnet
    const swapPrice = await cdp.evm.getSwapPrice({
      network: "ethereum",
      toToken: toToken.address as Address,
      fromToken: fromToken.address as Address,
      fromAmount: parseEther("0.1"), // 0.1 WETH in wei
      taker: ownerAccount.address,
    });

    // Check for liquidity
    if (!swapPrice.liquidityAvailable) {
      console.log("Insufficient liquidity available for this swap.");
      return;
    }

    // At this point we know liquidityAvailable is true and we can access all properties
    console.log("\nSwap Price:");
    console.log("-------------");
    console.log(`To Token Amount: ${formatUnits(swapPrice.toAmount, toToken.decimals)} ${toToken.symbol}`);
    console.log(`Min To Token Amount: ${formatUnits(swapPrice.minToAmount, toToken.decimals)} ${toToken.symbol}`);
    console.log(`From Token Amount: ${formatUnits(swapPrice.fromAmount, fromToken.decimals)} ${fromToken.symbol}`);
    console.log(`Gas Estimate: ${swapPrice.gas}`);
    
    // Calculate and display price ratios
    const fromAmountBigInt = BigInt(swapPrice.fromAmount);
    const toAmountBigInt = BigInt(swapPrice.toAmount);
    const minToAmountBigInt = BigInt(swapPrice.minToAmount);
    
    // Calculate exchange rate: How many toTokens per 1 fromToken
    const fromToToRate = Number(toAmountBigInt) / (10 ** toToken.decimals) * 
                         (10 ** fromToken.decimals) / Number(fromAmountBigInt);

    // Calculate minimum exchange rate with slippage applied
    const minFromToToRate = Number(minToAmountBigInt) / (10 ** toToken.decimals) * 
                           (10 ** fromToken.decimals) / Number(fromAmountBigInt);

    // Calculate maximum toToken to fromToken ratio with slippage
    const maxToToFromRate = Number(fromAmountBigInt) / (10 ** fromToken.decimals) *
                           (10 ** toToken.decimals) / Number(minToAmountBigInt);

    // Calculate exchange rate: How many fromTokens per 1 toToken
    const toToFromRate = Number(fromAmountBigInt) / (10 ** fromToken.decimals) *
                         (10 ** toToken.decimals) / Number(toAmountBigInt);
    
    console.log("\nToken Price Calculations:");
    console.log("------------------------");
    console.log(`1 ${fromToken.symbol} = ${fromToToRate.toFixed(toToken.decimals)} ${toToken.symbol}`);
    console.log(`1 ${toToken.symbol} = ${toToFromRate.toFixed(fromToken.decimals)} ${fromToken.symbol}`);

    // Calculate effective exchange rate with slippage applied
    console.log("\nWith Slippage Applied (Worst Case):");
    console.log("----------------------------------");
    console.log(`1 ${fromToken.symbol} = ${minFromToToRate.toFixed(toToken.decimals)} ${toToken.symbol} (minimum)`);
    console.log(`1 ${toToken.symbol} = ${maxToToFromRate.toFixed(fromToken.decimals)} ${fromToken.symbol} (maximum)`);
    console.log(`Maximum price impact: ${((fromToToRate - minFromToToRate) / fromToToRate * 100).toFixed(2)}%`);

    // Check for any issues
    if (swapPrice.issues.allowance) {
      console.log("\nAllowance Issues:");
      console.log(`Current Allowance: ${swapPrice.issues.allowance.currentAllowance}`);
      console.log(`Spender: ${swapPrice.issues.allowance.spender}`);
    }
    
    if (swapPrice.issues.balance) {
      console.log("\nBalance Issues:");
      console.log(`Current Balance: ${swapPrice.issues.balance.currentBalance}`);
      console.log(`Required Balance: ${swapPrice.issues.balance.requiredBalance}`);
    }

    if (swapPrice.issues.simulationIncomplete) {
      console.log("\n⚠️ WARNING: Simulation incomplete. Results may not be accurate.");
    }
  } catch (error) {
    console.error("Error getting swap price:", error);
  }
}

main().catch(console.error); 

File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/python/evm/swaps/get_swap_price.py
# Usage: uv run python evm/swaps/get_swap_price.py
#!/usr/bin/env python3
# Usage: uv run python evm/get_swap_price.py

"""
Example: Get Swap Price

This example demonstrates how to get a price quote for swapping tokens
without executing the swap. This is useful for displaying estimated
outputs to users before they commit to a swap.
"""

import asyncio
from decimal import Decimal

from cdp import CdpClient
from dotenv import load_dotenv

load_dotenv()


async def main():
    """Get a swap price quote."""
    async with CdpClient() as cdp:
        # Get or create an account to use as the taker
        account = await cdp.evm.get_or_create_account(name="PriceCheckAccount")
        print(f"Using account: {account.address}\n")
        
        # Token addresses on Base
        USDC = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
        WETH = "0x4200000000000000000000000000000000000006"
        
        # Get a price quote for swapping 100 USDC to WETH
        print("Getting swap price quote...")
        print(f"From: 100 USDC")
        print(f"To: WETH")
        print(f"Network: Base\n")
        
        try:
            # Get the price quote
            priceQuote = await cdp.evm.get_swap_price(
                from_token=USDC,
                to_token=WETH,
                from_amount="100000000",  # 100 USDC (6 decimals)
                network="base",
                taker=account.address  # Address where the from_token balance is located
            )
            
            # Display the price quote details
            print("📊 Price Quote:")
            print(f"   Quote ID: {priceQuote.quote_id}")
            print(f"   From: {Decimal(priceQuote.from_amount) / Decimal(10**6):.2f} USDC")
            print(f"   To: {Decimal(priceQuote.to_amount) / Decimal(10**18):.6f} WETH")
            
            # Calculate and display the price
            from_amount_decimal = Decimal(priceQuote.from_amount) / Decimal(10**6)
            to_amount_decimal = Decimal(priceQuote.to_amount) / Decimal(10**18)
            price_per_usdc = to_amount_decimal / from_amount_decimal
            
            print(f"   Price: 1 USDC = {price_per_usdc:.8f} WETH")
            print(f"   Expires at: {priceQuote.expires_at}")
            
        except Exception as e:
            print(f"❌ Error getting quote: {e}")


if __name__ == "__main__":
    asyncio.run(main()) 